designs/2020-timing-list-size/README.md

```diff
@@ -0,0 +1,245 @@
+no-irregular-whitespace |    43.847 |     5.1%
+valid-jsdoc             |    40.346 |     4.7%
+handle-callback-err     |    39.153 |     4.6%
+space-infix-ops         |    35.444 |     4.1%
+no-undefined            |    25.693 |     3.0%
+no-shadow               |    22.759 |     2.7%
+no-empty-class          |    21.976 |     2.6%
+semi                    |    19.359 |     2.3%
+```
+
+<!-- One-paragraph explanation of the feature. -->
+
+## Motivation
+
+Optimizing linting performance can become important in a large codebase or with particularly expensive rules. In order to determine what rules to focus on for performance optimizations, it's necessary to measure the running times of various rules while running linting on a codebase.
+
+While the current output of performance metrics for the top 10 longest-running rules helps, it doesn't give the complete picture, especially when there may be hundreds of rules enabled (from ESLint core and third-party plugins).
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+## Detailed Design
+
+We propose reusing the existing the `TIMING` environment variable to also specify the desired list size.
+
+- The performance list will continue to show only when the `TIMING` variable is present (no change).
+- When the list shows, we will read the integer value of the `TIMING` variable to determine the number of rules to show in the list.
+- A minimum list size of 10 will be used to maintain backwards compatability (and because there's not much reason to shorten the list beyond that anyway).
+- A special string value of `all` will remove the limit entirely (and make it clear that the user does not want a limit).
+
+| Command | Behavior |
+| --- | --- |
+| `eslint` | do not show |
+| `TIMING=0 eslint` | show the first 10 rules (due to minimum of 10) |
+| `TIMING=1 eslint` | show the first 10 rules (due to minimum of 10) |
+| `TIMING=5 eslint` | show the first 10 rules (due to minimum of 10) |
+| `TIMING=10 eslint` | show the first 10 rules |
+| `TIMING=11 eslint` | show the first 11 rules |
+| `TIMING=15 eslint` | show the first 15 rules |
+| `TIMING=100 eslint` | show the first 100 rules |
+| `TIMING=all eslint` | show all the rules |
+
+This is a concise, easy-to-type means of gaining additional configurability without adding additional environment variables.
+
+To implement this, we will add a function to calculate the list size in [TIMING.js](https://github.com/eslint/eslint/blob/master/lib/linter/timing.js):
+
+```js
+/**
+ * Decide how many rules to show in the output list.
+ * @returns {number} the number of rules to show
+ */
+function getListSize() {
+    if (process.env.TIMING === "all") {
+        return Number.POSITIVE_INFINITY;
+    }
+
+    const MINIMUM_SIZE = 10;
+    const TIMING_ENV_VAR_AS_INTEGER = Number.parseInt(process.env.TIMING, 10);
+
+    return Math.max(MINIMUM_SIZE, TIMING_ENV_VAR_AS_INTEGER);
```

Comment:
For `TIMING=true`, this function will return `NaN` and `display` will show nothing as `slice` returns `[]` if the second argument was `NaN`.

Reply:
Thanks, I fixed for the code for this case. Will also add test cases in the actual PR.